### PR TITLE
[serve] Add basic upscaling test using cluster_utils

### PR DESCRIPTION
**Pull request by edoakes**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/tests/test_cluster.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    # Add a node with another CPU, the final replica should get placed
+    # and the deploy goal should be done.
+    cluster.add_node(num_cpus=1)
+    assert client._wait_for_goal(goal_ref)
+    pids3 = get_pids(3)
+    assert pids2.issubset(pids3)
+
+
+@pytest.mark.skip("Currently hangs due to max_task_retries=-1.")
```
Comment (edoakes) [inline python/ray/serve/tests/test_cluster.py:64]: @simon-mo this is actually quite bad. Right now the handle doesn't gracefully handle replica failures at all; if a request is enqueued after a replica crashed, the request will hang until the replica is brought back up.

Comment (edoakes) [reply python/ray/serve/tests/test_cluster.py:64]: I think what we should be doing is not using ray's max_restarts or max_task_retries and instead just restart replicas ourselves

Comment (edoakes) [reply python/ray/serve/tests/test_cluster.py:64]: Implemented here: https://github.com/ray-project/ray/pull/15047

**Review by architkulkarni — APPROVED** [review]
Looks great!  This is a useful test to have.

File: python/ray/serve/tests/test_cluster.py (hunk 1 above)
Comment (simon-mo) [reply python/ray/serve/tests/test_cluster.py:64]: agreed. it's easier to perform app level healthcheck and restart for backends. we should still use ray's fault tolerance mechanism for controller though

**Review by simon-mo — APPROVED** [review]

**Pull request merged**